runtime/lib/capi/ExecutionContext.hpp

```diff
@@ -59,6 +60,50 @@ class MemoryManager final {
     bool contains(void *ptr) { return _impl.contains(ptr); }
 };
 
+class SharedLibraryManager final {
+  private:
+    void *_handler{NULL};
+
+  public:
+    SharedLibraryManager() {}
+
+    SharedLibraryManager(std::string filename)
+    {
+        if (filename.find(".so") == std::string::npos) {
+            return;
+        }
+        _handler = dlopen(filename.c_str(), RTLD_LAZY | RTLD_DEEPBIND);
+        if (!_handler) {
+            char *error_msg = dlerror();
+            throw RuntimeException(error_msg);
+        }
+    }
+
+    ~SharedLibraryManager()
+    {
+        // dlclose returns non-zero on error
+        // but we cannot throw exceptions in a destructor
+        // TODO: How to handle this case?
```

Comment:
dlclose just doesn't fit into the C++ resource allocation model.. We can either ignore errors and pretend that the library is closed (or issue a warning) or call dlclose somewhere else. I can think of keeping a singleton list of all to-be-closed libs and processing it during the last step of the quantum device execution workflow. This destructor then would just add its library into this list.

Reply:
At the moment, each SharedLibraryManager increases and decreases the reference of the library they open. I think this is fine for now.